File: apple/taOSc/PairingGrantView.swift
import SwiftUI

final class PairingGrantViewModel: ObservableObject {
    @Published var phase: Phase
    private let baseURL: URL
    private let displayName: String
    private var pollTask: Task<Void, Never>?

    enum Phase: Equatable {
        case creating
        case pending(verifyCode: String)
        case approved
        case denied
        case expired
        case unreachable(String)
    }

    init(baseURL: URL, displayName: String) {
        self.baseURL = baseURL
        self.displayName = displayName
        self.phase = .creating
    }

    func start() {
        guard phase == .creating else { return }
        pollTask?.cancel()
        pollTask = Task { @MainActor in
            do {
                let response = try await PairingService.createPairRequest(
                    baseURL: baseURL,
                    platform: "ios",
                    displayName: displayName
                )
                self.phase = .pending(verifyCode: response.verify_code)
                await poll(requestId: response.pair_request_id)
            } catch {
                self.phase = .unreachable(error.localizedDescription)
            }
        }
    }

    func cancel() {
        pollTask?.cancel()
    }

    deinit {
        pollTask?.cancel()
    }

    private func poll(requestId: String) async {
        while !Task.isCancelled {
            do {
                let status = try await PairingService.pollPairRequest(
                    baseURL: baseURL,
                    requestId: requestId
                )
                switch status {
                case .pending:
                    try? await Task.sleep(nanoseconds: 3_000_000_000)
                case .approved(let token):
                    try? KeychainStore.shared.saveToken(token)
                    self.phase = .approved
                    return
                case .denied:
                    self.phase = .denied
                    return
                case .expired:
                    self.phase = .expired
                    return
                }
            } catch {
                self.phase = .unreachable(error.localizedDescription)
                return
            }
        }
    }
}

struct PairingGrantView: View {
    @StateObject private var viewModel: PairingGrantViewModel
    let onComplete: () -> Void
    let onDismiss: () -> Void

    init(baseURL: URL, displayName: String, onComplete: @escaping () -> Void, onDismiss: @escaping () -> Void) {
        _viewModel = StateObject(wrappedValue: PairingGrantViewModel(
            baseURL: baseURL,
            displayName: displayName
        ))
        self.onComplete = onComplete
        self.onDismiss = onDismiss
    }

    var body: some View {
        Group {
            switch viewModel.phase {
            case .creating:
                ProgressView("Creating pairing request...")
                    .onAppear {
                        viewModel.start()
                    }

            case .pending(let verifyCode):
                PendingGrantView(verifyCode: verifyCode)

            case .approved:
                ProgressView("Pairing approved. Opening...")
                    .onAppear {
                        onComplete()
                    }

            case .denied:
                TerminalStateView(
                    icon: "xmark.circle.fill",
                    iconColor: .red,
                    title: "Pairing Denied",
                    message: "The request was denied. Please try again.",
                    buttonTitle: "Back",
                    action: { onDismiss() }
                )

            case .expired:
                TerminalStateView(
                    icon: "clock.badge.exclamationmark.fill",
                    iconColor: .orange,
                    title: "Pairing Expired",
                    message: "The verification code has expired. Please try again.",
                    buttonTitle: "Back",
                    action: { onDismiss() }
                )

            case .unreachable(let message):
                ErrorRetryView(message: message) {
                    viewModel.phase = .creating
                    viewModel.start()
                }
            }
        }
        .navigationTitle("Pairing")
        .navigationBarTitleDisplayMode(.inline)
    }
}

struct PendingGrantView: View {
    let verifyCode: String

    var body: some View {
        VStack(spacing: 24) {
            Text("Compare this code with the one shown on your taOS instance.")
                .font(.subheadline)
                .foregroundColor(.secondary)
                .multilineTextAlignment(.center)
                .padding(.horizontal)

            Text(verifyCode)
                .font(.system(.title, design: .monospaced))
                .fontWeight(.bold)
                .tracking(8)
                .padding()
                .frame(maxWidth: .infinity)
                .background(.ultraThinMaterial)
                .cornerRadius(12)
                .padding(.horizontal)

            ProgressView("Waiting for approval...")
                .padding()
        }
        .padding()
    }
}

struct TerminalStateView: View {
    let icon: String
    let iconColor: Color
    let title: String
    let message: String
    let buttonTitle: String
    let action: () -> Void

    var body: some View {
        VStack(spacing: 24) {
            Image(systemName: icon)
                .font(.system(size: 60))
                .foregroundColor(iconColor)

            Text(title)
                .font(.title2)
                .fontWeight(.semibold)

            Text(message)
                .font(.subheadline)
                .foregroundColor(.secondary)
                .multilineTextAlignment(.center)
                .padding(.horizontal)

            Button(buttonTitle) {
                action()
            }
            .buttonStyle(.borderedProminent)
        }
        .padding()
    }
}

struct ErrorRetryView: View {
    let message: String
    let retry: () -> Void

    var body: some View {
        VStack(spacing: 24) {
            Image(systemName: "exclamationmark.octagon")
                .font(.system(size: 60))
                .foregroundColor(.red)

            Text("Server Unreachable")
                .font(.title2)
                .fontWeight(.semibold)

            Text(message)
                .font(.subheadline)
                .foregroundColor(.secondary)
                .multilineTextAlignment(.center)
                .padding(.horizontal)

            Button("Retry") {
                retry()
            }
            .buttonStyle(.borderedProminent)
        }
        .padding()
    }
}

File: android/app/src/main/java/com/taosc/taosc/PairingGrantView.kt
package com.taosc.taosc

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.Button
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.delay

sealed class PairingPhase {
    data object Creating : PairingPhase()
    data class Pending(val verifyCode: String) : PairingPhase()
    data object Approved : PairingPhase()
    data object Denied : PairingPhase()
    data object Expired : PairingPhase()
    data class Unreachable(val message: String) : PairingPhase()
}

@Composable
fun PairingGrantScreen(
    baseUrl: String,
    onComplete: () -> Unit,
    onDismiss: () -> Unit,
    credentialStore: CredentialStore,
    settingsStore: SettingsStore
) {
    var phase by remember { mutableStateOf<PairingPhase>(PairingPhase.Creating) }
    var retryToken by remember { mutableStateOf(0) }

    LaunchedEffect(baseUrl, retryToken) {
        try {
            val service = PairingService()
            val response = service.createPairRequest(baseUrl, "android", "Android Device")
            phase = PairingPhase.Pending(response.verifyCode)

            while (true) {
                val status = service.pollPairRequest(baseUrl, response.pairRequestId)
                when (status) {
                    PairRequestStatus.Pending -> delay(3000)
                    is PairRequestStatus.Approved -> {
                        credentialStore.saveToken(status.scopedToken)
                        phase = PairingPhase.Approved
                        onComplete()
                        return@LaunchedEffect
                    }
                    PairRequestStatus.Denied -> {
                        phase = PairingPhase.Denied
                        return@LaunchedEffect
                    }
                    PairRequestStatus.Expired -> {
                        phase = PairingPhase.Expired
                        return@LaunchedEffect
                    }
                }
            }
        } catch (e: CancellationException) {
            throw e
        } catch (e: Exception) {
            phase = PairingPhase.Unreachable(e.message ?: "Unknown error")
        }
    }

    when (val currentPhase = phase) {
        is PairingPhase.Creating -> {
            Column(
                modifier = Modifier.fillMaxSize(),
                verticalArrangement = Arrangement.Center,
                horizontalAlignment = Alignment.CenterHorizontally
            ) {
                Text("Creating pairing request...")
                CircularProgressIndicator(modifier = Modifier.padding(top = 16.dp))
            }
        }
        is PairingPhase.Pending -> {
            Column(
                modifier = Modifier.fillMaxSize(),
                verticalArrangement = Arrangement.Center,
                horizontalAlignment = Alignment.CenterHorizontally
            ) {
                Text(
                    text = "Compare this code with the one shown on your taOS instance.",
                    modifier = Modifier.padding(horizontal = 32.dp)
                )
                Text(
                    text = currentPhase.verifyCode,
                    modifier = Modifier.padding(16.dp)
                )
                CircularProgressIndicator(modifier = Modifier.padding(top = 16.dp))
            }
        }
        is PairingPhase.Approved -> {
            Column(
                modifier = Modifier.fillMaxSize(),
                verticalArrangement = Arrangement.Center,
                horizontalAlignment = Alignment.CenterHorizontally
            ) {
                Text("Pairing approved. Opening...")
                CircularProgressIndicator(modifier = Modifier.padding(top = 16.dp))
            }
        }
        is PairingPhase.Denied -> {
            TerminalStateView(
                title = "Pairing Denied",
                message = "The request was denied. Please try again.",
                buttonTitle = "Back",
                action = onDismiss
            )
        }
        is PairingPhase.Expired -> {
            TerminalStateView(
                title = "Pairing Expired",
                message = "The verification code has expired. Please try again.",
                buttonTitle = "Back",
                action = onDismiss
            )
        }
        is PairingPhase.Unreachable -> {
            ErrorRetryView(
                message = currentPhase.message,
                retry = {
                    phase = PairingPhase.Creating
                    retryToken++
                }
            )
        }
    }
}

@Composable
fun TerminalStateView(
    title: String,
    message: String,
    buttonTitle: String,
    action: () -> Unit
) {
    Column(
        modifier = Modifier.fillMaxSize(),
        verticalArrangement = Arrangement.Center,
        horizontalAlignment = Alignment.CenterHorizontally
    ) {
        Text(text = title, modifier = Modifier.padding(bottom = 8.dp))
        Text(
            text = message,
            modifier = Modifier.padding(horizontal = 32.dp, vertical = 8.dp)
        )
        Button(
            onClick = action,
            modifier = Modifier.padding(top = 16.dp)
        ) {
            Text(buttonTitle)
        }
    }
}

@Composable
fun ErrorRetryView(
    message: String,
    retry: () -> Unit
) {
    Column(
        modifier = Modifier.fillMaxSize(),
        verticalArrangement = Arrangement.Center,
        horizontalAlignment = Alignment.CenterHorizontally
    ) {
        Text(text = "Server Unreachable", modifier = Modifier.padding(bottom = 8.dp))
        Text(
            text = message,
            modifier = Modifier.padding(horizontal = 32.dp, vertical = 8.dp)
        )
        Button(
            onClick = retry,
            modifier = Modifier.padding(top = 16.dp)
        ) {
            Text("Retry")
        }
    }
}
